### PR TITLE
Fixing non-NaNable ccd images

### LINKED ISSUE #157: Setting saturated pixels to `NaN` fails if image is integer
While testing some of the changes with the new photometry settings I was using an uncalibrated image. `single_image_photometetry` failed on [this line](https://github.com/feder-observatory/stellarphot/blob/main/stellarphot/photometry/photometry.py#L259) with 

```python-traceback
File ~/Development/astronomy/stellarphot/stellarphot/photometry/photometry.py:222, in single_image_photometry(ccd_image, sourcelist, camera, observatory_location, aperture_radius, inner_annulus, outer_annulus, shift_tolerance, max_adu, fwhm_estimate, use_coordinates, include_dig_noise, reject_too_close, reject_background_outliers, passband_map, fwhm_by_fit, fname, logline)
    219     return None, None
    221 # Set high pixels to NaN
--> 222 ccd_image.data[ccd_image.data > max_adu] = np.nan
    224 # Extract necessary values from sourcelist structure
    225 star_ids = sourcelist['star_id'].value

ValueError: cannot convert float NaN to integer
```

I think the fix can be just converting the data to `float`.

### PR DESCRIPTION
**Pull request by JuanCab**
I implemented a fix for the ccddata not being NaN-able by converting it to float if needed.  Can you check that your image will now process.

I also updated the CHANGES.rst file to cover recent changes.

Fixes #157

### CODE REVIEW

File: stellarphot/photometry/photometry.py [hunk 1]
```diff
@@ -255,7 +255,9 @@ def single_image_photometry(ccd_image, sourcelist, camera, observatory_location,
                "SKIPPING THIS IMAGE!")
         return None, None
 
-    # Set high pixels to NaN
+    # Set high pixels to NaN (make sure ccd_image.data is a float array first)
+    if ccd_image.data.dtype != float:
```
Comment (mwcraig) [inline stellarphot/photometry/photometry.py:259]: I think you can just cast it as a `float` without checking the type first. If it is already `float` the cast just doesn't do anything. (Also, use `isinstance` for type checks :))

**Review by mwcraig — CHANGES_REQUESTED** [review]

File: stellarphot/photometry/photometry.py (hunk 1 above)
Comment (JuanCab) [reply stellarphot/photometry/photometry.py]: Fixed.

File: stellarphot/photometry/photometry.py [hunk 2]
```diff
@@ -255,8 +255,8 @@ def single_image_photometry(ccd_image, sourcelist, camera, observatory_location,
                "SKIPPING THIS IMAGE!")
         return None, None
 
-    # Set high pixels to NaN
-    ccd_image.data[ccd_image.data > max_adu] = np.nan
+    # Set high pixels to NaN (make sure ccd_image.data is a float array first)
```
Comment (mwcraig) [inline stellarphot/photometry/photometry.py:258]: I think you still need a line here making the data float, it just doesn't need a type check:

```suggestion
    # Set high pixels to NaN (make sure ccd_image.data is a float array first)
    ccd_image.data = ccd_image.data.astype(float)
```

**Review by mwcraig — CHANGES_REQUESTED** [review]

File: stellarphot/photometry/photometry.py (hunk 2 above)
Comment (JuanCab) [reply stellarphot/photometry/photometry.py:258]: Didn't I cast it to float during the NaN replacement? (On iPhone, can't see code)

Comment (mwcraig) [reply stellarphot/photometry/photometry.py:258]: You cast it as float for the comparison (inside the brackets) but the data itself is still int I think

**Comment by JuanCab** [comment]
OK, it is now fixed.  Sigh.  I would have sworn I had been clever.  Can you PLEASE test this with the integer image, just in case?

Juan


On Aug 17, 2023, at 3:35 PM, Matt Craig ***@***.***> wrote:

@mwcraig commented on this pull request.

________________________________

In stellarphot/photometry/photometry.py<https://github.com/feder-observatory/stellarphot/pull/158#discussion_r1297712936>:

> @@ -255,8 +255,8 @@ def single_image_photometry(ccd_image, sourcelist, camera, observatory_location,
                "SKIPPING THIS IMAGE!")
         return None, None

-    # Set high pixels to NaN
-    ccd_image.data[ccd_image.data > max_adu] = np.nan
+    # Set high pixels to NaN (make sure ccd_image.data is a float array first)


You cast it as float for the comparison (inside the brackets) but the data itself is still int I think

File: stellarphot/photometry/photometry.py (hunk 2 above)
Comment (JuanCab) [reply stellarphot/photometry/photometry.py:258]: I think I fixed it now.

**Comment by mwcraig** [comment]
I just pushed a test which I verified fails without your fix 🎉 but unfortunately found out about #161 along the way 😭 .

**Review by mwcraig — APPROVED** [review]

**Pull request merged**